docker_parallel_runner.py

```diff
@@ -0,0 +1,320 @@
+
+    @property
+    def docker_compose_project_name(self) -> str:
+        m = re.search(r"^([a-z0-9][a-z0-9_-]*).*", os.path.basename(self.suite_file))
+        if m is None:
+            raise ValueError(
+                f"Invalid test suite: {self.suite_file}. "
+                "The path's basename should conform to COMPOSE_PROJECT_NAME format. "
+                "See https://docs.docker.com/compose/project-name/#set-a-project-name")
+        return m.group(1)
+
+    @property
+    def cpus_needed(self) -> float:
+        # What uses CPU in a test suite?
+        # - RP broker containers
+        # - client containers
+        # - control container (there is always just 1)
+        # This formula assumes 1 CPU per container
+        return float(self.rp_cluster_size + self.client_count + 1)
+
+    def __str__(self):
+        return repr(self)
+
+
+def generate_ssh_key():
+    # ssh key will be used to connect to the control container
+    # add a timeout - since ssh-keygen can hang waiting for user input (e.g. "Overwrite (y/n)?")
+    if os.path.exists("id_ed25519"):
+        logging.info("SSH key already exists. Skipping generation.")
+        return
+    cmd = ["ssh-keygen", "-t", "ed25519", "-f", "id_ed25519", "-N", ""]
+    logging.info(f"Running command: {' '.join(shlex.quote(arg) for arg in cmd)}")
+    subprocess.check_call(cmd, timeout=10)
+
+
+class SuiteRunner:
+    def __init__(self,
+                 rebuild_script: str,
+                 up_script: str,
+                 ready_script: str,
+                 down_script: str,
+                 fetch_logs_script: str,
+                 test_suite_script: str,
+                 docker_compose_project_name: str):
+        self._rebuild_script = rebuild_script
+        self._up_script = up_script
+        self._ready_script = ready_script
+        self._down_script = down_script
+        self._fetch_logs_script = fetch_logs_script
+        self._test_suite_script = test_suite_script
+        self._docker_compose_project_name = docker_compose_project_name
+
+    def _run(self, cmd: List[str], extra_env: dict = {}):
+        logging.info(f"Running command: {' '.join(shlex.quote(arg) for arg in cmd)}")
+        if "COMPOSE_PROJECT_NAME" in os.environ:
+            raise EnvironmentError(
+                "COMPOSE_PROJECT_NAME is already set in initial environment - but this script will set its own!")
+        subprocess.check_call(cmd, env={
+            **os.environ,
+            **extra_env,
+            "COMPOSE_PROJECT_NAME":
+                self._docker_compose_project_name})
+
+    def rebuild(self):
+        # Rebuild scripts by default will generate an SSH key at a fixed common location
+        # We want to skip this step, as rebuild may be happening across concurrent suites.
+        # Instead we will prepare the SSH key once before starting any suites.
+        return self._run([self._rebuild_script],
+                         extra_env={"SKIP_SSH_KEYGEN": "1"})
```

Comment:
could you add retries to guard against network errors here? retries currently work at taskfile level (ie `rebuild.sh` are being retried), which I think won't work as is now

Reply:
Note the retry policy does not change *at all* in this PR.  The entrypoint is STILL `test.all.sh` which as you noted is being retried at the taskfile level.

To improve things future we should add granular retries *within* `rebuild.sh`.  E.g. rebuild does a lot of things, only some of can fail for retryable reasons.  E.g. if docker build `apt install` is failing for retryable error, we should make the `apt` retry.

May be we can do this in a follow up / TODO - not directly related to this PR (entrypoint remains the same `test.all.sh` etc.).

Reply:
I don't think that retries will work. we were using an arbitrary exit code to retry on (that was `10`) and taskfile was running retries only if the task exited with `10` to skip retrying the entire script

Reply:
Got you, I'll post a PR to propagate non-zero exit codes.  That would be the cleanest way for now.